File: eventmanager/classtier.ts
// API-facing skill class tier type. This is a plain string-literal union rather
// than the Prisma-generated `ClassTier` enum: Encore's API schema parser cannot
// use Prisma's runtime enum object as a type ("value used as type"). The values
// are byte-for-byte identical to the Prisma enum, so they interoperate with the
// database layer without conversion.
export type ClassTier = "PRE_OP" | "OP" | "G3" | "G2" | "G1";

// Class tiers ordered from lowest to highest skill (issue #3). PRE-OP is the
// entry tier; G1 is the top grade, mirroring Netkeiba-style grading.
export const CLASS_TIER_ORDER: ClassTier[] = ["PRE_OP", "OP", "G3", "G2", "G1"];

// Human-readable labels. The enum identifiers cannot contain hyphens, so
// PRE_OP is displayed as "PRE-OP".
export const CLASS_TIER_LABELS: Record<ClassTier, string> = {
  PRE_OP: "PRE-OP",
  OP: "OP",
  G3: "G3",
  G2: "G2",
  G1: "G1",
};

// Class restrictions exist so participants are matched with others of equal
// skill.
// Eligibility rules:
// - If there is no restriction, any participant is allowed.
// - Same class: allowed.
// - One class lower: allowed.
// - Two or more classes lower: rejected.
// - `PRE_OP`: allowed only for `PRE_OP` and `OP`.
export function isEligible(
  participantTier: ClassTier | null,
  eventRestriction: ClassTier | null,
): boolean {
  // PRE_OP and OP are treated as equivalent to unrestricted / none (null)
  const normParticipant = (participantTier === "PRE_OP" || participantTier === "OP") ? null : participantTier;
  const normRestriction = (eventRestriction === "PRE_OP" || eventRestriction === "OP") ? null : eventRestriction;

  if (normRestriction === null) {
    return true;
  }
  if (normParticipant === null) {
    return false;
  }

  if (normParticipant === normRestriction) {
    return true;
  }

  const pIdx = CLASS_TIER_ORDER.indexOf(normParticipant);
  const rIdx = CLASS_TIER_ORDER.indexOf(normRestriction);

  if (pIdx === -1 || rIdx === -1) {
    return false;
  }

  // Other racers can enter their own class or one class lower.
  // Since CLASS_TIER_ORDER is ["PRE_OP", "OP", "G3", "G2", "G1"],
  // one class lower restriction means rIdx === pIdx - 1.
  return rIdx === pIdx - 1;
}

File: eventmanager/classtier.go
package eventmanager

// Skill class tiers, mirroring Netkeiba-style grading (issue #3).
//
// Mirrors ts-legacy/eventmanager/classtier.ts. Values are byte-for-byte
// identical to the Prisma ClassTier enum.
type ClassTier string

const (
	ClassTierPreOp ClassTier = "PRE_OP"
	ClassTierOp    ClassTier = "OP"
	ClassTierG3    ClassTier = "G3"
	ClassTierG2    ClassTier = "G2"
	ClassTierG1    ClassTier = "G1"
)

// ClassTierOrder lists tiers from lowest to highest skill.
var ClassTierOrder = []ClassTier{
	ClassTierPreOp,
	ClassTierOp,
	ClassTierG3,
	ClassTierG2,
	ClassTierG1,
}

// ClassTierLabels holds human-readable labels (PRE_OP displays as "PRE-OP").
var ClassTierLabels = map[ClassTier]string{
	ClassTierPreOp: "PRE-OP",
	ClassTierOp:    "OP",
	ClassTierG3:    "G3",
	ClassTierG2:    "G2",
	ClassTierG1:    "G1",
}

func tierIndex(t ClassTier) int {
	for i, v := range ClassTierOrder {
		if v == t {
			return i
		}
	}
	return -1
}

// IsEligible reports whether a participant tier may enter an event with the
// given class restriction. Nil means unrestricted/unclassified.
//
// Eligibility rules (mirrors ts-legacy/eventmanager/classtier.ts isEligible):
//   - No restriction: anyone allowed.
//   - Same class: allowed.
//   - One class lower restriction: allowed.
//   - PRE_OP and OP normalize to unrestricted.
func IsEligible(participantTier, eventRestriction *ClassTier) bool {
	norm := func(t *ClassTier) *ClassTier {
		if t == nil {
			return nil
		}
		if *t == ClassTierPreOp || *t == ClassTierOp {
			return nil
		}
		return t
	}
	p, r := norm(participantTier), norm(eventRestriction)

	if r == nil {
		return true
	}
	if p == nil {
		return false
	}
	if *p == *r {
		return true
	}
	pIdx, rIdx := tierIndex(*p), tierIndex(*r)
	if pIdx == -1 || rIdx == -1 {
		return false
	}
	return rIdx == pIdx-1
}
